DS-4061 expose sitemaps and schedule automatic updates

benbosman opened the pull request:
## References
* Fixes #2802. 
* Link to (original) [JIRA](https://jira.lyrasis.org/browse/DS-4061) ticket.  Also fixes [DS-1901](https://jira.lyrasis.org/browse/DS-1901)
* The REST contract was considered not relevant

## Description
This will return the contents of a given sitemap, typically as HTML or XML. Sample URLs
 * /server/sitemaps/sitemap_index.html contains a list of links to other maps (links to Server) with content type HTML
 * /server/sitemaps/sitemap0.html contains a list of links to items (links to Angular) with content type HTML
 * /server/sitemaps/sitemap_index.xml contains an XML with links to other maps (links to Server) with content type XML
 * /server/sitemaps/sitemap0.xml contains an XML with links to items (links to Angular) with content type XML

## Instructions for Reviewers
Make sure that the config of the server and Angular are correct
Make sure that **after** deploying this PR, you use `dspace/bin/dspace generate-sitemaps` to create the sitemaps with the correct URLs
Open /server/sitemaps/sitemap_index.html or /server/sitemaps/sitemap_index.xml and you can follow links

As discussed, this PR also contains a configurable solution to automatically rebuild sitemaps without a cron job. This can be found at https://github.com/DSpace/DSpace/pull/2858/files#diff-540baf79d64235945c1d30aff36ee09eR15

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

MarieVerdonck committed:
71734: Sitemap REST controller

MarieVerdonck committed:
71736: Changes to sitemap contents

MarieVerdonck committed:
71737: Sitemap REST IT

MarieVerdonck committed:
71738: Automatic cron job for sitemaps

MarieVerdonck committed:
71859: Endpoint path sitemaps changed

MarieVerdonck committed:
Merge branch 'dspace-origin-master' into w2p-71734_sitemaps

# Conflicts:
#	dspace-server-webapp/src/test/java/org/dspace/app/rest/RootRestResourceControllerIT.java
#	dspace-server-webapp/src/test/java/org/dspace/app/rest/ShibbolethRestControllerIT.java

tdonohue commented:
Closing & reopening the PR to try and ping Travis to build this PR.  For some strange reason it stalled & didn't build this PR.

tdonohue commented on dspace/config/spring/api/task-scheduler.xml:
```diff
@@ -0,0 +1,20 @@
+
+  <!-- Configure parameters -->
+  <task:scheduled-tasks scheduler="myScheduler">
+    <task:scheduled ref="GenerateSitemaps" method="generateSitemapsScheduled" cron="0 15 1 * * ?"/>
```
This needs to be configurable, as we shouldn't assume that everyone needs this to run at the same rate.  We also should document the syntax expected here -- it looks like the syntax is defined here? https://www.quartz-scheduler.org/api/2.3.0/org/quartz/CronTrigger.html

Maybe we should create a config in `dspace.cfg` named `sitemap.cron` and changing this entry to say `cron="${sitemap.cron}"`?  
```
# Define cron for how frequently the sitemap should refresh.
# Defaults to running daily at 1:15am
# Syntax is defined at https://www.quartz-scheduler.org/api/2.3.0/org/quartz/CronTrigger.html
sitemap.cron = 0 15 1 * * ?
```

Or, if we feel a cron syntax is too complex, we could switch this to using `fixed-rate` and have a simple configuration like:
```
# Refresh sitemap every 4 hours (4 x 60 x 60 x 1,000ms)
sitemap.refreshrate = 28800000
```

I don't feel strongly here, but I suspect users might have an easier time configuring the `refreshrate` instead of a custom cron schedule.

abollini replied:
it would be also useful to include the ping to the search engine as a configurable option similar to how now can be specified with the `-p` from the command line.

benbosman replied:
I did assume that spring config sufficed for it to be configurable
Moving it to dspace.cfg makes it easier to override it in local.cfg, but there are many things configured in spring which are more likely to be adjusted but only possible in the Spring config
Is there a guideline about which spring config is handled in dspace.cfg?

tdonohue replied:
@benbosman : generally, anything that is more likely to need customization (and a cronjob type of schedule would) I tend to recommend moving to DSpace configuration, so that it can be updated in `local.cfg`.  Spring Configuration is OK for *advanced* configurations which we don't want most sites to modify.

In this scenario, I do think the sitemap generation schedule will often need customization (based on size of site and other cronjobs that folks might have running).  

That said, I don't feel that we need to make the `-p` option configurable.  Instead, I'd recommend we just make the default here to use the `-a` option, which will automatically ping any search engines configured in `sitemap.engineurls`...so that is already easily configurable.

benbosman replied:
This has been adjusted

tdonohue reviewed (changes requested):
@benbosman : I haven't tested this fully yet, but have done a code review.  Overall it looks good, but I think a few things here need to be made configurable (see comments inline), namely the path of sitemaps and the update schedule.

I also hadn't realized that these task schedulers were so easy to do in Spring.  Seems like a great opportunity for future PRs to move more cron-based tasks into the scheduler.

**UPDATE:** I've tested this, and it seems to work (at least by manually running `dspace generate-sitemaps`).  It took me a moment to realize I had to go to `sitemap_index.html`, rather than just `sitemap.html`, but that's fine.  I *did* however notice we may need to update the links between the index files & map files...as I think they will eventually need to appear like UI links/paths (once the UI piece is built out).  So, for example: http://localhost:8080/server/sitemaps/sitemap_index.xml has a `<loc>http://localhost:8080/server/sitemaps/sitemap0.xml</loc>`, which is valid for "http://localhost:8080/server", but wouldn't be valid for "http://localhost:4000/".  But, that's something we may need to find a way to cleanup later on.

mwoodiupui commented:
Internal task scheduler:  eww, please no.  As a sysadmin. I intensely dislike having to keep track of applications' private schedulers.  Each supported OS has a standard task scheduler of its own that the admin. already is aware of and understands.  Let's use them.

tdonohue commented:
@mwoodiupui : The reason we need an internal scheduler in this scenario is that it's been proven (by Google Scholar) that *most* DSpace sites are failing to configure the cron job that enables SiteMaps.  This has a major effect on DSpace's out-of-the-box SEO. See https://jira.lyrasis.org/browse/DS-1901

So, conceptually, I agree with you.  But, in this scenario, we need a way to enable SiteMaps by default and "schedule" them to auto-update (they'd be too time consuming to dynamically generate, so we need a backend process to auto-update them on a regular, but configurable, schedule -- likely at least once per day).  So, that's exactly why we are looking at Spring Tasks for this use case, as it's a simple way to do this without building something new ourselves.

mwoodiupui commented:
@tdonohue If Scholar is okay with occasional partial/broken sitemaps because I took the system down in the middle of an update, not remembering that we have magic scheduled jobs hidden inside DSpace, then I guess I withdraw my objection in this specific case.

tdonohue commented:
@mwoodiupui : If this automation results in partial/broken sitemaps, then maybe that's something we simply need to fix in the process, i.e. "don't replace existing sitemap files until process fully completes".  That'd be the ideal here.  But, in any case, Scholar's main complaint is that they find the _majority_ of DSpace sites they index have either no sitemaps or very outdated ones (i.e. they ran the sitemap generator once, or it used to work, but now their cron job is failing & they didn't notice).  So, I'm told that Scholar has come to find they cannot trust DSpace's sitemaps at all, and therefore they mostly ignore them -- which is a bad practice for SEO, and exactly why Scholar keeps asking us to fix DSpace's sitemap generation process to automate it.

abollini commented on dspace/config/spring/api/task-scheduler.xml:
```diff
@@ -0,0 +1,20 @@
+    http://www.springframework.org/task/spring-task.xsd">
+
+  <!-- Configure the scheduler -->
+  <task:scheduler id="myScheduler" pool-size="10"/>
```
can we found a way to make easier turn off the scheduler without touching the xml file? an option could be our own implementation wrapping the default one with an additional boolean flag, if turned off the scheduler will just silent.
This is needed in deployment where there are multiple nodes, off course we must have the scheduler running only on a node (where we can set the property to true using ENV variable)

benbosman replied:
That seems to be quite some overhead to avoid changing the config
Removing the scheduler would achieve the same result

abollini replied:
it would be hard to change the xml file just on a single node than set an ENV variable, if this is too complex to do in this PR I would like at least to have a ticket to keep track of that. We should start a section in the documentation to explain how to deploy dspace in cluster and point to this and other features that would require the selection of a "main" node

tdonohue replied:
For what it's worth, it from this answer it *sounds like* setting the `cron = -` will disable the schedule: https://stackoverflow.com/a/57655646

So, this is even more argument for moving the `cron` to our Configuration files...as that'd let you easily change it or enable/disable the task via your `local.cfg`, or environment variables (as our ConfigurationService also reads env)

tdonohue replied:
It's also possible to enable/disable by configurations + annotations like this: https://stackoverflow.com/a/51198831  That also should be possible in DSpace, as Spring annotations can read from our configuration.  For example, we already use `@ConditionalOnProperty` to do things like enable/disable OAI-PMH via an `oai.enabled` configuration: https://github.com/DSpace/DSpace/blob/main/dspace-oai/src/main/java/org/dspace/xoai/controller/DSpaceOAIDataProvider.java#L56

abollini replied:
both options look fine to me (thanks @tdonohue for the investigation). I slightly prefer the last ConditionalOnProperty over the first but both are fine

benbosman replied:
@tdonohue Marie has verified the options in detail.
The `cron = -` syntax is not supported (it's not in the JavaDocs, and testing it results in `Cron expression must consist of 6 fields (found 1 in "-")`
The `@ConditionalOnProperty` is also not supported, most likely it can only be used if everything is moved to Java.

I don't see any options to configure this in local.cfg without bigger Java changes, which may be out-of-scope if it only has the goal to ease overriding config

abollini commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/SitemapRestController.java:
```diff
@@ -0,0 +1,148 @@
+                sender.withDisposition(MultipartFileSender.CONTENT_DISPOSITION_ATTACHMENT);
+            }
+
+            Context context = ContextUtil.obtainContext(request);
+
+            // We have all the data we need, close the connection to the database so that it doesn't stay open during
+            // download/streaming
+            context.complete();
```
I think that @benbosman is right here, forcing the context to complete will free up resource earlier that we will never use. The context.complete method is already internally protected with an `isValid()` check we should eventually put a guard against NPE but this would be a very odd scenario so I'm also fine in just having the NPE and throw a 500

tdonohue replied:
Ok, sounds fine. I'll close this request.

abollini reviewed (changes requested):
I asked for some additional test related to the security risk to expose server file with this dedicated endpoint (we should be protected by the spring firewall but better to have a test in place) and for an improvement to allow to turn off the internal scheduler if needed (sysadmin preference or multinodes deployment)

benbosman commented on dspace-api/src/main/java/org/dspace/app/sitemap/GenerateSitemaps.java:
```diff
@@ -168,21 +176,19 @@ public static void generateSitemaps(boolean makeHTMLMap,
 
         for (Community comm : comms) {
-            String url = handleURLStem + comm.getHandle();
+            String url = uiURLStem + "/communities/" + comm.getID();
```
At this point in time, they don't work. This can be reverted if the handles work, but I'd prefer to have a working solution
I also thought there were plans to remove the requirement to have a handle (so it would become optional)

tdonohue replied:
@benbosman : aha, ok. I didn't realize these were not working.  I'm OK with this change, I was mostly asking the reason why it was made.  I do feel handles will be made optional eventually, so you are correct this is a step in that direction.  I'll close out this request as resolved. Thanks!

benbosman commented:
@tdonohue I have googled for requirements stating the sitemap should be available on the same hostname as the UI, but didn't find any. As soon as this branch and the robots.txt work has been completed, this can be put to a test by having the demo repository indexed.

@mwoodiupui I agree with a preference to use the OS crontab, and we'll most likely disable the automatic cron job for our clients. But as stated by Tim, there are many people who don't use the OS crontab, and just don't generate sitemaps

I've answered some questions inline and scheduled changes for the other feedback

MarieVerdonck committed:
72351: sitemaps path to config + in tests &
script extension to delete sitemaps dir > called at end of tests &
test for malicious file traversal

abollini reviewed (changes requested):
I have provided some additional feedback about the initial remediation introduced.
My understanding is that you are going to process the feedback provided here https://github.com/DSpace/DSpace/pull/2858/files#r459460371 in subsequent commits

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/SitemapRestControllerIT.java:
```diff
@@ -0,0 +1,141 @@
+            .withIssueDate("2015-8-3")
+            .build();
+
+        runDSpaceScript("generate-sitemaps");
```
@abollini : This looks to be resolved already.  See the new `destroy()` method below this one.  Before/After annotations were also added.

abollini replied:
thanks to point me to that I haven't noted it. I have added another minor suggestion, not blocking but I think would make the code cleaner (and faster) https://github.com/DSpace/DSpace/pull/2858#discussion_r463247105

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/SitemapRestControllerIT.java:
```diff
@@ -0,0 +1,167 @@
+                   .andExpect(status().isNotFound());
+    }
+
+    @Test(expected = ServletException.class)
+    public void testSitemap_fileSystemTraversal_dspaceCfg() throws Exception {
+        //** WHEN **
+        //We attempt to use endpoint for malicious file system traversal
+        getClient().perform(get("/" + SITEMAPS_ENDPOINT + "/%2e%2e/config/dspace.cfg"));
+    }
+
+    @Test(expected = ServletException.class)
+    public void testSitemap_fileSystemTraversal_dspaceCfg2() throws Exception {
+        //** WHEN **
+        //We attempt to use endpoint for malicious file system traversal
+        getClient().perform(get("/" + SITEMAPS_ENDPOINT + "/%2e%2e%2fconfig%2fdspace.cfg"));
+    }
```
I agree, this seems like it should be throwing a `ResourceNotFoundException` (or maybe responding with a simple 404 response code, i.e. `.andExpect(status().isNotFound());` ) based on the code in SitemapRestController here: https://github.com/DSpace/DSpace/pull/2858/files#diff-be5a61ded7d68e1169e3c3576ca9ce30R93

If that's not possible (for some reason), we could leave this as-is & add either a TODO or bug ticket. As-is, this test does prove that you cannot access other files from this endpoint, but the `@Test(expected=ServletException.class)` just seems like the wrong response.

In debugging this, we could add an additional simplistic test here to ensure an invalid filename (something like "notasitemap.html") also throws a `ResourceNotFoundException` or 404 response.

benbosman replied:
This is valid behavior because it's an attempt to bypass security. The message returned is also: `The request was rejected because the URL contained a potentially malicious String "%2e"`
For non-attacks, just incorrect filenames, 404 is returned instead

tdonohue replied:
I'm OK with this for now. If we later decide to improve on it, we can do so. The bigger point is that this sort of request should throw an error & it obviously does. Thanks @benbosman

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/SitemapRestControllerIT.java:
```diff
@@ -0,0 +1,167 @@
+    @After
+    public void destroy() throws Exception {
+        // delete sitemaps generated by tests in before
+        runDSpaceScript("generate-sitemaps", "-d");
```
can we call the static method directly here? it would be more easy to understand and efficient

benbosman replied:
Do you have an example of the alternative method you're asking for? I didn't see any problems with using the DSpace script

abollini replied:
I simply mean to call directly `org.dspace.app.sitemap.GenerateSitemaps.deleteSitemaps()` as it is a static public method

benbosman commented on dspace-api/src/main/java/org/dspace/app/sitemap/GenerateSitemaps.java:
```diff
@@ -61,6 +61,8 @@
     private static final ConfigurationService configurationService =
         DSpaceServicesFactory.getInstance().getConfigurationService();
 
+    public static final String SITEMAPS_ENDPOINT = "/sitemaps";
```
This has been updated in https://github.com/DSpace/DSpace/pull/2858/commits/1ba6a748874dbf7b3952ed0cb393a796e2202180

benbosman commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/SitemapRestController.java:
```diff
@@ -0,0 +1,148 @@
+ * @author Maria Verdonck (Atmire) on 08/07/2020
+ */
+@Controller
+@RequestMapping("/sitemaps")
```
This has been updated in https://github.com/DSpace/DSpace/pull/2858/commits/1ba6a748874dbf7b3952ed0cb393a796e2202180

benbosman commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/SitemapRestControllerIT.java:
```diff
@@ -0,0 +1,141 @@
+    @Autowired
+    ConfigurationService configurationService;
+
+    private final static String SITEMAPS_ENDPOINT = "/sitemaps";
```
This has been updated in https://github.com/DSpace/DSpace/pull/2858/commits/1ba6a748874dbf7b3952ed0cb393a796e2202180

benbosman commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/SitemapRestControllerIT.java:
```diff
@@ -0,0 +1,141 @@
+        getClient().perform(get(SITEMAPS_ENDPOINT + "/notValidSiteMapFile"))
+                   //** THEN **
+                   .andExpect(status().isNotFound());
+    }
```
This has been updated in https://github.com/DSpace/DSpace/pull/2858/commits/1ba6a748874dbf7b3952ed0cb393a796e2202180

MarieVerdonck committed:
72426: Sitemap cron schedule configurable in dspace.cfg

MarieVerdonck committed:
Revert whitespace changes

MarieVerdonck committed:
Merge branch 'dspace-origin-master' into w2p-71734_sitemaps

MarieVerdonck committed:
poolsize fix

MarieVerdonck committed:
Collection and Community Builder moved fix

MarieVerdonck committed:
71738: Sitemap scheduler with annotations

tdonohue commented on dspace/config/dspace.cfg:
```diff
@@ -1429,6 +1429,9 @@ webui.content_disposition_threshold = 8388608
 sitemap.dir = ${dspace.dir}/sitemaps
 
+# sitemaps endpoint path used in SitemapRestController and in the url paths when sitemaps are generated
+sitemap.path = sitemaps
```
Minor thing here. I'd recommend we comment this out by default and change comment slightly. So, update to something like:

```
# Customize the path of sitemaps in the server webapp
# Defaults to "sitemaps", which means the sitemaps are available at ${dspace.server.url}/sitemaps
# sitemap.path = sitemaps
```

tdonohue commented on dspace/config/dspace.cfg:
```diff
@@ -1442,6 +1445,12 @@ sitemap.engineurls = http://www.google.com/webmasters/sitemaps/ping?sitemap=
 
+# Define cron for how frequently the sitemap should refresh.
+# Defaults to running daily at 1:15am
+# Remove (comment) out this config to disable the sitemap scheduler
```
Another very minor change.  Please update this line to also describe how to disable via local.cfg.  I've also found (via testing) that this config is not refreshed automatically (which isn't a big deal), so we should document that you need to restart Tomcat.

So, let's say something like:
```
# Remove (comment out) this config to disable the sitemap scheduler.
# Sitemap scheduler can also be disabled by setting to "-" (single dash) in local.cfg.
# Keep in mind, changing the schedule requires rebooting your servlet container, e.g. Tomcat.
```

tdonohue reviewed (approved):
👍 I've reviewed & tested this again today, and it now looks great to me.  Thanks @benbosman and @MarieVerdonck for switching this over into Spring Annotations! I tested this by setting `sitemap.cron = 0 * 16 * * ?` (recreates Sitemaps every minute in the 16:00 hour) and waited one minute to find the sitemaps created in the `[dspace]/sitemaps/` folder.

I've added a few minor suggestions on improving the comments in `dspace.cfg`.  But, once those are added, I think the rest of this PR looks good as-is.  Thanks again!

MarieVerdonck committed:
72505: sitemaps config feedback applied & changed base url of sitemap_index files to ui.url

tdonohue commented on dspace-api/src/main/java/org/dspace/app/sitemap/GenerateSitemaps.java:
```diff
@@ -150,14 +180,10 @@ public static void main(String[] args) throws Exception {
+    public static void generateSitemaps(boolean makeHTMLMap, boolean makeSitemapOrg) throws SQLException, IOException {
+        String sitemapsEndpoint = configurationService.getProperty("sitemap.path", "sitemaps");
+        String uiURLStem = configurationService.getProperty("dspace.ui.url");
+        String sitemapStem = uiURLStem + "/" + sitemapsEndpoint + "/sitemap";
```
Minor change required here.  Based on updates to https://github.com/DSpace/dspace-angular/pull/825 , this `sitemapStem` should be updated to just be:
`String sitemapStem = uiURLStem + "/sitemap";`

The reason is that, at the UI level, the `sitemapsEndpoint` has no meaning. It's only valid for the REST API.  This is needed, to ensure that the sitemap files are available at the root level of the UI (e.g. http://localhost:400/sitemap_index.xml , http://localhost:400/sitemap0.xml)

MarieVerdonck replied:
@tdonohue Applied fix; https://github.com/DSpace/DSpace/pull/2858/commits/4b86f331d43041d3991124534c4c546068d75546

tdonohue replied:
Thanks @MarieVerdonck !  I'll give this one last test then and merge it assuming it all works (and it should).

tdonohue reviewed (commented):
@benbosman : Very minor request for an update to this PR.  After testing this again with the updated Angular PR (https://github.com/DSpace/dspace-angular/pull/825), I've realized that the /sitemap0.xml (and similar files) appear under the wrong UI path. I've added an inline comment to describe the minor change.

Beyond this minor bug fix, I'm +1 this PR.

abollini commented:
I'm happy with the changes in https://github.com/DSpace/DSpace/pull/2858/commits/b4e64e5bf373ee2be6cb304f88b0177ba131c4b1 I have provided an answer here https://github.com/DSpace/DSpace/pull/2858#discussion_r473965164 for the other minor changes that I have suggested but it is not a blocking one so in any case, once the bug found by @tdonohue is solved I'm +1 with that

MarieVerdonck committed:
72505: Url sitemaps fix

MarieVerdonck committed:
72505: Url sitemaps change - test fix

tdonohue commented:
I've retested this with the Angular UI `robots.txt` (which was already merged on `main`) and it works after @MarieVerdonck 's minor updates (see above comments).

As this is at +2, I'm merging immediately.  Thanks again @benbosman and @MarieVerdonck !
